src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1756,11 +1762,10 @@ private BaseListSyntax ParseBaseList(SyntaxToken typeKeyword, bool haveParameter
                 list.Add(argumentList is object ? _syntaxFactory.PrimaryConstructorBaseType(firstType, argumentList) : (BaseTypeSyntax)_syntaxFactory.SimpleBaseType(firstType));
 
                 // any additional types
-                while (true)
+                int lastTokenPosition = -1;
+                while (IsMakingProgress(ref lastTokenPosition))
                 {
-                    if (this.CurrentToken.Kind == SyntaxKind.OpenBraceToken ||
-                        this.CurrentToken.Kind == SyntaxKind.SemicolonToken ||
-                        this.IsCurrentTokenWhereOfConstraintClause())
```

Comment:
Note: I intentionally simplified `this.IsCurrentTokenWhereOfConstraintClause()` to `this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword`, to align the abort logic used before and after the success case.